File: source/yeon/5주차/code/1182부분수열의합.cpp
// 백준 1182 부분수열의 합
// cpp solved by combinations and brute-force
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;
vector<int> arr; // save input from user
vector<vector<int>> result; // save combination result
vector<int> sub;
int N, S;

void combination(int k){
    // if index is same as size of arr
    // push it to result[]
    if(k == N){
        result.push_back(sub);
    }else{
        sub.push_back(arr[k]);
        combination(k+1);
        sub.pop_back();
        combination(k+1);
    }
}

int main(void){
    int num, cnt = 0;
    long long sum;
    
    scanf("%d %d",&N, &S);
    for (int i = 0; i < N; i++){
        scanf("%d", &num);
        arr.push_back(num);
    }

    // start recurisve from index 0 
    combination(0);

    // check is value which prevent counting cnt 
    // even though there's nothing to calculate 
    int check;
    for(int i = 0; i <result.size();i++){
        sum = 0, check = 0;
        for(int j = 0; j < result[i].size();j++){
            sum += result[i][j];
            check++;
        }
        if(sum == S && check > 0) cnt++;
    }
    printf("%d", cnt);
    return 0;
}

File: source/yeon/5주차/code/1182부분수열의합.py
# 백준 1182 부분수열의 합
# python solved by combinations and brute-force
from itertools import combinations

N, S = map(int, input().split())
arr = list(map(int,input().split()))

cnt = 0
# make combinations 1 to N
# and check the sum of each case 
# if there's any answer exist 
for i in range(1,N+1):
    a = list(combinations(arr,i))
    for j in a :
        sum = 0
        for h in range(i):
            sum += j[h]
        if sum == S :
            cnt +=1
print(cnt)